CMakeLists.txt

```diff
@@ -110,152 +110,42 @@ set(AZURE_STORAGE_LITE_SOURCE
-    if(NOT((DEFINED CURL_INCLUDE_DIRS) OR (DEFINED CURL_LIBRARIES)))
-      find_package(CURL REQUIRED)
-    endif()
-  endif()
-
-  #Constructing OPENSSL path on windows.
-  if((DEFINED OPENSSL_ROOT_DIR) AND (NOT (OPENSSL_ROOT_DIR STREQUAL "")))
-    if(CMAKE_CL_64)
-      set(INHERIT_ENV "x64")
-    else()
-      set(INHERIT_ENV "x86")
-    endif()
-
-    if(OPENSSL_LINK_TYPE STREQUAL "static")
-      add_definitions(-DOPENSSL_STATICLIB)
-      set(OPENSSL_LIBRARY_NAME "ssleay32.lib")
-      set(OPENSSL_CRYPTO_LIBRARY_NAME "libeay32.lib")
-    else()
-      set(OPENSSL_LIBRARY_NAME "ssleay32.dll")
-      set(OPENSSL_CRYPTO_LIBRARY_NAME "libeay32.dll")
-    endif()
-
-    set(OPENSSL_SSL_LIBRARY "${OPENSSL_ROOT_DIR}\\lib\\${CMAKE_BUILD_TYPE}\\${OPENSSL_LIBRARY_NAME}")
-    set(OPENSSL_CRYPTO_LIBRARY "${OPENSSL_ROOT_DIR}\\lib\\${CMAKE_BUILD_TYPE}\\${OPENSSL_CRYPTO_LIBRARY_NAME}")
-    set(OPENSSL_INCLUDE_DIR "${OPENSSL_ROOT_DIR}\\include\\openssl\\")
-    message (STATUS "Constructed openssl library path: ${OPENSSL_SSL_LIBRARY}")
-    message (STATUS "Constructed openssl include path: ${OPENSSL_INCLUDE_DIR}")
-    message (STATUS "Constructed openssl crypto library path: ${OPENSSL_CRYPTO_LIBRARY}")
-    unset(INHERIT_ENV)
-    unset(OPENSSL_LIBRARY_NAME)
-    unset(OPENSSL_CRYPTO_LIBRARY_NAME)
-  else()
-    if(OPENSSL_LINK_TYPE STREQUAL "static")
-      add_definitions(-DOPENSSL_STATICLIB)
-    endif()
+add_library(azure-storage-lite ${AZURE_STORAGE_LITE_HEADER} ${AZURE_STORAGE_LITE_SOURCE})
 
-    #if there isn't a customized openssl library and include, then use FindOpenSSL.cmake to find Openssl on Windows
-    if(NOT((DEFINED OPENSSL_INCLUDE_DIR) AND (DEFINED OPENSSL_CRYPTO_LIBRARY) AND (DEFINED OPENSSL_SSL_LIBRARY)))
-      message(STATUS "OpenSSL folder is not specified, need to find it automatically.")
-      find_package(OpenSSL REQUIRED)
-    endif()
-  endif()
+option(USE_OPENSSL "Use OpenSSL instead of GnuTLS" ON)
 
-  if(OPENSSL_LINK_TYPE STREQUAL "static")
-    add_definitions(-DOPENSSL_STATICLIB)
-  endif()
-  add_definitions(-DUSE_OPENSSL)
-  set(OPENSSL_MSVC_STATIC_RT "true")
-  if (BUILD_WITH_MT)
-    set(CMAKE_CXX_FLAGS_RELEASE "/MT")
-    set(CMAKE_CXX_FLAGS_DEBUG "/MT")
-  endif()
-  #adding extra dependency libraries for windows
-  set(EXTRA_LIBRARIES ${EXTRA_LIBRARIES} rpcrt4 ws2_32 wldap32 crypt32 normaliz)
-  include_directories(${PROJECT_SOURCE_DIR}/include ${CURL_INCLUDE_DIRS} ${OPENSSL_INCLUDE_DIR} ${EXTRA_INCLUDE})
+set(CMAKE_THREAD_PREFER_PTHREAD TRUE)
+set(THREADS_PREFER_PTHREAD_FLAG TRUE)
+find_package(Threads REQUIRED)
+find_package(CURL REQUIRED)
+if(USE_OPENSSL)
+  target_compile_definitions(azure-storage-lite PRIVATE -DUSE_OPENSSL)
+  find_package(OpenSSL REQUIRED)
+  list(APPEND EXTRA_LIBRARIES OpenSSL::SSL OpenSSL::Crypto)
+else()
+  find_package(GnuTLS  REQUIRED)
+  list(APPEND EXTRA_INCLUDE_DIRS ${GNUTLS_INCLUDE_DIR})
+  list(APPEND EXTRA_LIBRARIES ${GNUTLS_LIBRARIES})
 endif()
```

Comment:
Can Mac also use GnuTLS? Is this validated?

Reply:
Yes, if installed via homebrew, it works.